### PR TITLE
Feature/network scanner control

### PR DESCRIPTION
**Pull request by amitschendel**
## Overview

___

## **Type**
enhancement, tests

___

## **Description**
- Implemented and tested network scanner functionality to identify unauthenticated services.
- Added rego support for network scanner integration.
- Updated dependencies to support new functionality.
- Registered new rego function for network scanning in the processor handler.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>networkscanner_test.go</strong><dd><code>Adding Network Scanner Functionality Test</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

core/pkg/opaprocessor/networkscanner_test.go
<li>Added a test for <code>isUnauthenticatedService</code> function.<br> <li> Demonstrates handling of unauthenticated services using a mock Redis <br>server.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/kubescape/pull/1592/files#diff-ef19a51820990ffc8a781004907f1d59b634754611d0118a8188adeddd4fa6b1">+78/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>networkscanner.go</strong><dd><code>Implement Network Scanner Functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

core/pkg/opaprocessor/networkscanner.go
<li>Implemented <code>isUnauthenticatedService</code> to check for unauthenticated <br>services.<br> <li> Utilizes <code>kubescape-network-scanner</code> for scanning.<br> <li> Handles timeout with a log error.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/kubescape/pull/1592/files#diff-73cf4bcbd3694319e90677b67b80248cf66121230075b582bcb7b2b8036741ee">+49/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>utils.go</strong><dd><code>Rego Support for Network Scanner</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

core/pkg/opaprocessor/utils.go
<li>Added rego function declaration and definition for network scanner.<br> <li> Handles conversion of service, port, and namespace parameters for the <br>scanner.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/kubescape/pull/1592/files#diff-12815c2517c676451e92ef77704d81804d0cf4f4e54752f2e1268777ea574cc6">+30/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>processorhandler.go</strong><dd><code>Register Network Scanner Rego Function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

core/pkg/opaprocessor/processorhandler.go
- Registered the network scanner rego function.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/kubescape/pull/1592/files#diff-e9448838afe79a9ed0e110dd6a475c42005fadef0095a053c59e5fcf818c6fc3">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Dependencies</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>go.mod</strong><dd><code>Update Dependencies for Network Scanner</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

go.mod
<li>Updated several package versions including <code>github.com/google/uuid</code> and <br><code>go.opentelemetry.io/otel</code>.<br> <li> Added new dependencies for network scanning and other functionalities.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/kubescape/pull/1592/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6">+68/-33</a>&nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by matthyx** [comment]
can you rebase?

File: core/pkg/opaprocessor/networkscanner.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+const (
+	ServiceSuffix = ".svc.cluster.local"
+	Timeout       = time.Second * 3
+)
+
+// Check if the service is unauthenticated using kubescape-network-scanner.
+func isUnauthenticatedService(host string, port int, namespace string) bool {
+	// Skip kube-system namespace.
+	if namespace == "kube-system" {
```
Comment (dwertent) [inline core/pkg/opaprocessor/networkscanner.go:19]: @slashben Do we want to skip on the KS namespace as well?

File: core/pkg/opaprocessor/networkscanner.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	servicediscovery "github.com/kubescape/kubescape-network-scanner/cmd"
+)
+
+const (
+	ServiceSuffix = ".svc.cluster.local"
+	Timeout       = time.Second * 3
+)
+
+// Check if the service is unauthenticated using kubescape-network-scanner.
+func isUnauthenticatedService(host string, port int, namespace string) bool {
```
Comment (dwertent) [inline core/pkg/opaprocessor/networkscanner.go:17]: The function should receive the URL

**Comment by matthyx** [comment]
@amitschendel are we still considering this?

**Comment by amitschendel** [comment]
@matthyx I think this PR is not relevant anymore. 
@idohuber Is working on the new version.

**Pull request closed without merging**